File: src/main/java/com/todolist/task/service/exception/BusinessException.java
package com.todolist.task.service.exception;

import org.springframework.http.HttpStatus;

public class BusinessException extends RuntimeException {

  private final HttpStatus httpStatus;

  private final Error error;

  public BusinessException(Error error) {
    this(error, HttpStatus.BAD_REQUEST);
  }

  public BusinessException(Error error, HttpStatus httpStatus) {
    this(error, httpStatus, error.getValue());
  }

  protected BusinessException(Error error, HttpStatus httpStatus, String message) {
    super(message);
    this.error = error;
    this.httpStatus = httpStatus;
  }

  public Error getError() {
    return error;
  }

  public HttpStatus getHttpStatus() {
    return httpStatus;
  }
}

File: src/main/kotlin/com/todolist/task/service/exception/BusinessException.kt
package com.todolist.task.service.exception

import com.todolist.task.service.exception.ErrorMessage
import org.springframework.http.HttpStatus

class BusinessException : RuntimeException {

    val httpStatus: HttpStatus
    val errorMessage: ErrorMessage
    constructor(errorMessage: ErrorMessage) : this(errorMessage, HttpStatus.BAD_REQUEST)

    constructor(errorMessage: ErrorMessage, httpStatus: HttpStatus) : this(errorMessage, httpStatus, errorMessage.value)

    constructor(errorMessage: ErrorMessage, httpStatus: HttpStatus, message: String) : super(message) {
        this.errorMessage = errorMessage
        this.httpStatus = httpStatus
    }

}
